src/main/java/net/minecraftforge/common/brewing/MixingBrewingRecipe.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016-2021.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation version 2.1
+ * of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+package net.minecraftforge.common.brewing;
+
+import com.google.gson.JsonObject;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.item.alchemy.Potion;
+import net.minecraft.world.item.alchemy.PotionUtils;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.item.crafting.RecipeSerializer;
+import net.minecraft.world.item.crafting.ShapedRecipe;
+import net.minecraftforge.registries.ForgeRegistries;
+import net.minecraftforge.registries.ForgeRegistryEntry;
+import net.minecraftforge.registries.ObjectHolder;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public record MixingBrewingRecipe(@Nonnull ResourceLocation id,
+                                  @Nonnull Potion input,
+                                  @Nonnull Ingredient ingredient,
+                                  @Nonnull Potion output) implements IBrewingRecipe
+{
+    @ObjectHolder("forge:mixing_brewing")
+    public static final RecipeSerializer<MixingBrewingRecipe> SERIALIZER = null;
+
+    @Override
+    public boolean isInput(@Nonnull ItemStack stack)
+    {
+        final Item item = stack.getItem();
+        return (item == Items.POTION || item == Items.SPLASH_POTION || item == Items.LINGERING_POTION || item == Items.GLASS_BOTTLE) && PotionUtils.getPotion(stack) == input();
```

Comment:
Hardcoded entries makes it impossible to add new potion contains..

Reply:
MixingBrewingRecipe is intended to handle just the vanilla recipes
besides I don't know how else to check it because an instanceof check would be too broad (maybe a tag?)